src/metrics/ga4.py

```diff
@@ -0,0 +1,89 @@
+
+    # lsh@2023-02-13: there are no frames more complex than the 'prefix' anymore.
+    # pull any other logic in from ga3.py and dispatch as necessary.
+    ensure('prefix' in frame, "frame with id %r has no 'prefix' key. ga4 only supported on frames using a simple prefix." % frame['id'])
+    prefix = frame['prefix']
+
+    # cap the given date range (start_date, end_date) to frame boundaries if they extend beyond them.
+    start_date = frame['starts'] if start_date < frame['starts'] else start_date
+    end_date = frame['ends'] if end_date > frame['ends'] else end_date
+
+    # https://developers.google.com/analytics/devguides/reporting/data/v1
+    # https://developers.google.com/analytics/devguides/reporting/data/v1/api-schema
+    return {"dimensions": [{"name": "date"},
+                           {"name": "pagePathPlusQueryString"}],
+            "metrics": [{"name": "sessions"}],
+            # TODO: orderBys
+            "dateRanges": [{"startDate": ga_utils.ymd(start_date),
+                            "endDate": ga_utils.ymd(end_date)}],
+            "dimensionFilter": {
+                "filter": {
+                    "fieldName": "pagePathPlusQueryString",
+                    "stringFilter": {
+                        "matchType": "BEGINS_WITH",
+                        "value": prefix}}},
+            "orderBys": [
+                {"desc": True,
+                 "dimension": {"dimensionName": "date",
+                               "orderType": "NUMERIC"}}],
+            "limit": "10000"}
+
+def query_ga(ptype, query, replace_cache_files=False):
+    # urgh
+    start_dt = ga_utils.todt(query['dateRanges'][0]['startDate'])
+    end_dt = ga_utils.todt(query['dateRanges'][0]['endDate'])
+
+    results_type = ptype
+    results, output_path = ga_core.query_ga_write_results_v2(query, start_dt, end_dt, results_type)
+    return results
+
+# --- processing
+
+def prefixed_path_id(prefix, path):
+    path = urlparse(path).path.lower() # normalise
+    ensure(path.startswith(prefix), "path does not start with given prefix (%r): %s" % (prefix, path), ValueError)
+    path = path[len(prefix):].strip().strip('/') # /events => '', /events/foobar => foobar, /events/foo/bar/ => foo/bar
+    identifier = path.split('/', 1)[0] # '' => '', foobar => foobar, foo/bar => foo
+    return identifier
+
+def process_response(ptype, frame, response):
+    rows = response.get('rows')
+    if not rows:
+        LOG.warning("GA responded with no results", extra={'query': response['query'], 'ptype': ptype, 'frame': frame})
+        return []
+
+    id_fn = partial(prefixed_path_id, frame['prefix'])
+
+    def _process(row):
+        try:
+            datestr = row['dimensionValues'][0]['value']
+            path = row['dimensionValues'][1]['value']
+            count = row['metricValues'][0]['value']
+            identifier = id_fn(path)
+
+            if identifier is None:
```

Comment:
under what conditions does this happen? Should a warning be issued?

Reply:
it doesn't happen. it was copied from ga3 where some types of processors may have returned a None